Introduce TypeDB cluster support

farost opened the pull request:
## Usage and product changes

Introduce cluster support across all TypeDB 3.x driver languages (Rust, Java, Python, C#, C/C++, HTTP-TypeScript).

### gRPC

**This release contains breaking changes to public interfaces and will require minor code updates as a part of the migration.**

**This release is not compatible with the old GRPC protocol. The C headers have changed significantly, requiring updates in the dependent drivers.**

- Implement cluster-aware connection management. The driver discovers all replicas on connection open and automatically routes operations to the primary replica for strong consistency. You can specify multiple replicas of a cluster or just a single address -- even one replica is enough to get the whole information about the cluster and establish a consistent and automatically redirecting connection.
- Change `DriverOptions`. Add `DriverTlsConfig` for explicit TLS configuration: `disabled()`, `enabled_with_native_root_ca()`, `enabled_with_root_ca(path)`. Replaces the previous `is_tls_enabled` boolean.
- Add `ServerRouting` for controlling request routing:
  - `Auto` — driver selects the primary replica automatically, with failover on leadership changes. Currently, always strongly consistent.
  - `Direct(address)` — route to a specific replica. Useful for read-only workloads on secondaries where it's possible.
- Add `Server` API exposing replica metadata: `id`, `role` (Primary/Candidate/Secondary), `term`, `address`, and `version`.
- Add address translation for cloud/NAT deployments. Supports mapping public addresses to private connection addresses.
- Add primary failover with configurable retries (`DriverOptions::primary_failover_retries`, default: 1). On leadership change, the driver seeks the new primary and retries the operation.
- Refactor `Database` API: remove unused `ReplicaInfo` / `DatabaseReplicas`. Replica information is now server-level via `driver.servers()`.
- Add `request_timeout` to `DriverOptions` to ensure any operation of the driver (excluding operations within already opened transactions: queries, commits, ...) is executed within the set time limit. The driver returns an error if the time limit is reached without a response from a server. This prevents the request from hanging indefinitely if the server is online, but does not respond. 
- Rename `community` back to `core` to highlight the technical side of the edition used.
- Complete the format and the content of the API references.

### HTTP-TS

**The HTTP-TS driver IS backwards compatible. Updating it will not introduce build or runtime errors, but will let your applications support automatic redirects when putting multiple addresses in the constructor.** 

- Implement cluster-aware connection management. Unlike gRPC, its retry logic is not configurable, but follows a similar pattern: the driver tries executing a request against the first address specified in the constructor; in case of an explicit redirect in the error response it reauthenticates & redirects the request to the specified location; in case the error is preserved the driver tries the rest of the addresses in the list; if no addresses work, the request fails. 
- Add new messages for retrieving `Server` statuses.

## Implementation

### Rust driver

#### Server discovery and connection management

`ServerManager` orchestrates cluster topology. On connection open, the initial server responds with the full replica list (`ServerManager.All.Res`). The manager maintains:
- `replicas: RwLock<HashSet<AvailableServer>>` — discovered servers with replication status
- `replica_connections: RwLock<HashMap<Address, ServerConnection>>` — connection pool to each known replica
- `address_translation: RwLock<AddressTranslation>` — public-to-private address mapping for cloud deployments

Discovery happens at startup and is refreshed during failover via `refresh_replicas()`. New connections are created lazily, and stale connections (to removed replicas) are cleaned up.

#### Routing and failover

`ServerRouting::Auto` routes to the primary replica for strongly consistent operations. The primary is determined by the highest Raft term among replicas with `ReplicationRole::Primary`.

On failure, `execute_strongly_consistent()` distinguishes between:
- `ClusterReplicaNotPrimary` — leadership changed, seek new primary in existing replica set
- Network errors — the old primary is unreachable, seek primary excluding the failed address

Primary seeking (`seek_primary_replica_in`) connects to available replicas, calls `servers_all()` to get updated topology, and selects the primary with the highest term. Configurable via `DriverOptions::primary_failover_retries` (default: 1 retry).

#### Address translation

`Addresses` enum supports three modes:
- `Direct(Vec<Address>)` — single or multiple server addresses
- `Translated(HashMap<Address, Address>)` — public-to-private address mapping

Address translation enables cloud deployments where servers advertise public addresses but the driver connects via private addresses. The translation is applied bidirectionally: discovered public addresses are mapped to private addresses for connections, and private addresses are mapped back to public for display.

#### Server metadata

`AvailableServer` carries `ReplicationStatus` (id, role, term) from the Raft consensus layer. The `Replica` trait provides a uniform interface for accessing server metadata across the public and internal APIs.

`ServerVersion` stores the server distribution name and version string, retrieved via the `server_version` RPC.

#### TLS configuration

`DriverTlsConfig` replaces the previous boolean flag with explicit configuration:
- `disabled()` — plaintext, not recommended for production
- `enabled_with_native_root_ca()` — uses system trust roots (default)
- `enabled_with_root_ca(path)` — custom CA certificate for self-signed or internal CAs

### Language spefifics

All language drivers implement the same cluster-aware API:

**Java** (`java/`):
- `Driver` constructor accepts `DriverOptions` with TLS config and failover settings
- `ServerRouting` enum: `Auto` / `new Direct(address)`
- `Server` interface with `id()`, `role()`, `term()`, `address()`, `version()`
- `ReplicationRole` enum: `PRIMARY`, `CANDIDATE`, `SECONDARY`

**Python** (`python/`):
- `TypeDB.driver(addresses, credentials, driver_options)` factory
- `ServerRouting.Auto()` / `ServerRouting.Direct(address)`
- `Server` class with replication properties
- Moved `Database` and `DatabaseManager` to `typedb.api.database` package

**C#** (`csharp/`):
- `TypeDB.Driver(addresses, credentials, driverOptions)` constructor
- `ServerRouting.Auto` / `ServerRouting.Direct`
- `IServer` interface with `ReplicationRole` enum
- `DriverTlsConfig` with builder pattern

**C** (`c/`):
- New types: `Credential`, `DriverOptions`, `DriverTlsConfig`, `Server`, `ServerRouting`
- C API functions: `driver_new`, `driver_servers`, `server_get_address`, etc.
- SWIG bindings updated for Java and C# interop, fixed missing or excessive fields

**HTTP-TypeScript** (`http-ts/`):
- `Server` type with replication metadata
- `ServersListResponse` for the `/servers` endpoint

### Testing infrastructure

- BDD test steps updated for cluster scenarios across all languages
- Test encryption certificates regenerated
- `start-cluster-servers.sh` updated for new cluster configuration
- `tool/test/resources/config.yml` added for test server configuration

### Known limitations

- Blocking `thread::sleep` in async context during primary seeking (FIXME in `server_manager.rs:365`)

farost committed:
Add TypeDB Cluster support and failover (#765)

Introduce TypeDB Cluster support. Including:

Drivers can be created using a single address, multiple addresses, or
address translation (mapping between public user-facing addresses and
internal cluster addresses for special needs). Address translation can
be updated after the driver is created.

Server version and distribution, as well as the list of active replicas,
can be retrieved using TypeDB Driver.

Read operations can be executed using one of three strategies:
* Strong consistency level -- Strongest consistency, always up-to-date
due to the guarantee of the primary replica usage.
* Eventual consistency level -- Allow stale reads from any replica. May
not reflect latest writes (will not be supported in the first Cluster
release on the server side, so will probably return errors)
* Replica dependent consistency level -- The operation is executed
against the provided replica address only. Can be especially useful for
testing.

By default, all operations are executed using the strongest consistency
level. However, all read operations now support consistency level
specification. Read transactions can be configured through transaction
options.

Driver options can be used to configure the failover strategies used
with strong and eventual consistency levels.

`primary_failover_retries` -- Limits the number of attempts to redirect
a strongly consistent request to another primary replica in case of a
failure due to the change of replica roles. Defaults to 1.
`replica_discovery_attempts` -- Limits the number of driver attempts to
discover a single working replica to perform an operation in case of a
replica unavailability. Every replica is tested once, which means that
at most:
- {limit} operations are performed if the limit <= the number of
replicas.
- {number of replicas} operations are performed if the limit > the
number of replicas.
- {number of replicas} operations are performed if the limit is None.
Affects every eventually consistent operation, including redirect
failover, when the new primary replica is unknown. Defaults to None.

**Missing parts**:
* Address translation runtime update is not fully implemented, it
probably requires more updates, or can be removed for the first release.
* Testing cannot be automated without Cluster snapshots. Instead, for
local testing, use `temp-cluster-server` (instructions below).
* The driver works itself, the tests for Core pass. However, its
replication features are untested. There is a new `clustering`
integration test, which successfully boots up servers, but there are
some missing functionalities of the server that prevent the testing
progress. Once the PR referenced below is completed and merged, it can
be completed, as well. More BDD tests should be added here
https://github.com/typedb/typedb-behaviour/pull/373

This driver can be tested using the server from
https://github.com/typedb/typedb-cluster/pull/608. To do so, run `cargo
build` in `typedb-cluster`, and then copy the resulting binary to
`typedb-driver` like:
```
cp path-to-typedb/typedb-cluster/target/debug/typedb_server_bin path-to-typedb/typedb-driver/tool/test/temp-cluster-server/typedb
```
Then, test scripts like
`./tool/test/temp-cluster-server/start-cluster-servers.sh 3` can be
used.
Note that it is a temporary directory, created to use the server while
we are not able to use Cluster snapshots from Bazel.

Remove server_manager prints

Rustfmt and C impl cleanup

Update deps after rebase

Fix build

Fix encryption and C tests

Expose consistency too all the needed interfaces in Rust and C

Uncomment linux archives

Configure cluster tests better and introduce driver fields retrieval

Remove unready factory jobs and fix options interfaces

Add Java and Python BDD cluster init. Update artifacts

Temporarily make https/tls validation more friendly (#832)

Allow using `http` and `https` schemes in connection URLs, introducing a
temporary comparison of the specified scheme and the TLS driver option.
This is a temporary resolution until we decide how to make the GRPC
endpoint safe and easy to understand, yet comfortable for Cloud and
Cluster.

Additionally, introduce more Java interfaces as a part of its build fix.

Update TypeDB ref

Fix Java and Python build. Uncomment Java and Python CI. Update docs

Clean up build

Update version to 3.7.0-alpha-2. Fix test and update release notes

Fix python and windows c tests

Fix invalid test sizes

Fix python test rules

Fix tests

Update dependencies

Refactor DriverOptions tls configuration and preapre for 3.7.0-alpha-3

Fix build

Update dependencies

Update cluster artifact

Update cluster artifact

Rename community to core in internals and fix windows builds

Fix Rust doc generation

Update TypeDB artifact

Fix windows test job

Add GRPC requests timeouts (#840)

Add request timeouts for operations outside of transactions
(connections, users, databases, transaction opening). Now, instead of
hanging forever if the server does not respond, the operation fails with
an explicit error message on timeout.

These timeouts can be configured through driver options and are
defaulted to 2 hours. They do not affect transaction lifetimes, queries,
and commits (thus, commits can still be running forever -- an explicit
TODO is left for consideration).

This change replaces https://github.com/typedb/typedb-driver/pull/710

Use `tokio::time::timeout`.

Add request timeout todo

Rustfmt

Fix Java bdd tests

Remove large tests to limit the time given to the tests to 5 mins

Try stabilizing bdd tests in java

Remove bind address from connection addresses of clustering internal addresses

Rustfmt

Fix Javas consistency level constructors

Add cluster support BDD tests (#845)

Add cluster support BDD and fix uncovered bugs.

Affected drivers:
* rust
* java
* python
* ...

Extend cluster tests retries

farost committed:
Update Cargo files and docs after rebase

farost committed:
Fix build

farost committed:
Fix build again

farost committed:
Fix docs and update dependencies

farost committed:
Fix bdd tests

farost committed:
Try fixing tests

farost committed:
Don't consider AUT1 error as invalid token cred

farost committed:
Register AUT1 as an auth error

farost committed:
Prepare for release 3.8.0-alpha-1

farost committed:
Dissolve ConsistencyLevels and introduce ServerRouting (#851)

Refactor driver interfaces after alpha-stage testing:
* Remove `ConsistencyLevel`s due to the inability of the servers to
fully support such custom requests
* Instead, add more explicit `ServerRouting` to the requests that
support non-primary executions (server replica management for cluster
debugging)
* Cleanup naming (reduce the usage of the `Replica` keyword in favour of
pure `Server`)

Direct changes with the preservation of the old design.

farost committed:
Remove use_replication driver option (#852)

## Usage and product changes
Remove `use_replication` driver option as it is currently considered a
redundant functionality. It's okay if the driver tries connecting to all
the existing replicas, but then the needed requests can be redirected
using `ServerRouting`.

## Implementation

farost committed:
Enhance unknown server routing error

farost committed:
Update primary_failover_retries docs

farost committed:
Remove server discovery attempts options (#853)

## Usage and product changes
With the removal of the eventually consistent operations,
`server_discovery_attempts` driver option does not make much sense
anymore. Thus, it is being removed.

## Implementation

farost committed:
Cleanups after rebase

farost committed:
Update server artifacts

farost committed:
Add clustering support to C# (#856)

Extend and refactor the C# driver to support clustering. This includes:
* reformed driver constructors with explicit addresses;
* streamlined options APIs;
* server-related endpoints;
* server routing;
* usage of the updated C functions with automating routing support.

Mirrors the Java implementation.
Cleans up the dead code from both Java and C#.

farost committed:
Add dotnet ci jobs for the feature branch

farost committed:
Remove address translation update

farost committed:
Update proto and artifacts, the build will be broken

farost committed:
Replace replica management with the admin tool usage

farost committed:
Add request_timeout tests

farost committed:
Fix bdd steps and update artifacts

farost committed:
Update typedb artifacts

farost committed:
Fix HTTP interfaces for servers

farost commented on .circleci/windows/java/test_deploy_snapshot.bat:
```diff
@@ -25,7 +25,7 @@ powershell -Command "Move-Item -Path bazel-typedb-driver\external\typedb_artifac
 RD /S /Q typedb-all-windows
 powershell -Command "Move-Item -Path typedb-all-windows-* -Destination typedb-all-windows"
-START /B "" typedb-all-windows\typedb server --development-mode.enable=true
+START /B "" typedb-all-windows\typedb server --development-mode.enabled=true --server.http.enabled=false
```
bbbuuuuuugggggg

farost commented on .circleci/config.yml:
```diff
@@ -134,21 +134,21 @@ commands:
       - install-pip-requirements
       - run: |
-          tool/test/start-community-server.sh
+          tool/test/start-core-server.sh
```
`community` is very confusing, and we agreed that we use `core` when we talk about the repo, not the product.

farost commented on .circleci/config.yml:
```diff
@@ -793,7 +793,8 @@ jobs:
       - deploy-crate-release-unix
       - deploy-maven-release-unix
-      - deploy-http-ts-npm-release-unix
+# TODO: http-ts is disabled in the clustering branch
```
TODO: Uncomment

farost commented on .circleci/config.yml:
```diff
@@ -845,32 +846,32 @@ workflows:
           filters:
             branches:
-              only: [master]
+              only: [master, cluster-support-feature-branch]
```
TODO: remove these

farost commented on .factory/automation.yml:
```diff
@@ -104,49 +104,52 @@ build:
+        bazel run --config=ci @typedb_dependencies//tool/bazelinstall:remote_cache_setup.sh
+        bazel run --config=ci @typedb_dependencies//distribution/artifact:create-netrc
+ 
+        # TODO: run `bazel test --config=ci //rust/tests/behaviour/...` instead of multiple tests (cannot do Clustered migration now)
```
Will be available later in the future

farost replied:
I could add an ignore to the BDD steps declarations, but it would be unfair, because we still expect them to work as a requirement!

flyingsilverfin replied:
I dont mind our current setup of just running them one at a time wdyt?

farost replied:
This will be eliminated once database import is available in clusters. Since you don't mind the current set up, we'll leave it as is for now

farost commented on c/src/common/error.rs:
Refactoring the packaging in the `c` dir a little so it's easier to coordinate things

farost commented on c/src/common/mod.rs:
```diff
@@ -42,3 +46,10 @@ pub extern "C" fn string_iterator_next(it: *mut StringIterator) -> *mut c_char {
     free(it);
 }
+
+pub(crate) fn iterators_to_map<K: Eq + Hash, V>(
```
Address translation

farost commented on c/src/concept/concept.rs:
```diff
@@ -84,22 +76,6 @@ impl DatetimeAndTimeZone {
         }
     }
-
-    pub fn get_datetime_in_nanos(self) -> DatetimeInNanos {
```
SWIG generates its own thing

farost replied:
Although C getters might require these since the fields are not public... TODO check!

flyingsilverfin replied:
wait what do you mean - these are in the c API already?

farost replied:
All the C struct members are pub by default, so yes

farost commented on c/src/concept/mod.rs:
```diff
@@ -21,30 +21,15 @@ use std::ptr::addr_of_mut;
 
-/// Promise object representing the result of an asynchronous operation.
-/// Use \ref concept_promise_resolve(ConceptPromise*) to wait for and retrieve the resulting boolean value.
-pub struct ConceptPromise(BoxPromise<'static, Result<Option<Concept>>>);
```
Not used

farost commented on c/src/driver.rs:
```diff
@@ -0,0 +1,249 @@
+/// @param credentials The <code>Credentials</code> to connect with
+/// @param driver_options The <code>DriverOptions</code> to connect with
+#[no_mangle]
+pub extern "C" fn driver_new(
```
Renamed `driver_open` to `driver_new`. `open` is an outdated name from back when it was `connection_open`. This will break the other libs that already depend on our headers, but it's not the only breaking change, so I decided that it's time for a nice cleanup.

farost commented on c/swig/typedb_driver_csharp.swg:
```diff
@@ -150,17 +150,37 @@
 %noexception user_get_name;
 %noexception user_get_password_expiry_seconds;
 
-/* Replica info getters do not throw */
```
Counted, compared between csharp and java, removed the unused records, and ensured that every existing method is covered equally everywhere. Reordered so it's easier to compare all the alternatives in the future.

flyingsilverfin replied:
fantastic, thank you!

flyingsilverfin replied:
my goodness i hope we've not got any memory leaks anywhere. Actually, after this merges maybe we can write a little stress test in each driver that opens drivers and transaction, doing a small bulk load, to see if the memory just increases :?

farost replied:
It sounds like a very good idea, I've been striving for more stability in TypeDB and the tools. It's much easier with claude now. Created an issue: https://github.com/typedb/typedb-driver/issues/865

farost commented on c/swig/typedb_driver_java.swg:
```diff
@@ -439,8 +458,6 @@
 
 %promise(AnalyzedQuery, analyzed_query)
 %promise(QueryAnswer, query_answer)
-%promise(Concept, concept)
```
Removed

farost commented on c/swig/typedb_driver_java.swg:
```diff
@@ -439,8 +458,6 @@
 %promise(AnalyzedQuery, analyzed_query)
 %promise(QueryAnswer, query_answer)
-%promise(Concept, concept)
-%promise(Rule, rule)
```
Trash left from the 2.x days

farost commented on c/docs_structure.bzl:
```diff
@@ -21,127 +21,117 @@
-    "valuetype": "concept",
-
     # Analyze
+    "analyzed": "analyze",
```
Adjusted to the same ordering and typing in all the `x/docs_structure.bzl`

farost commented on csharp/Analyze/Function.cs:
```diff
@@ -24,6 +24,8 @@
 using TypeDB.Driver.Api.Analyze;
 using TypeDB.Driver.Common;
 
+using AnalyzeError = TypeDB.Driver.Common.Error.Analyze;
```
Mirrors the `java` impl. The previous implementors (claude) were lazy.

farost commented on csharp/Api/DriverTlsConfig.cs:
```diff
@@ -0,0 +1,124 @@
+        /// DriverTlsConfig tlsConfig = DriverTlsConfig.Disabled();
+        /// </code>
+        /// </example>
+        public static DriverTlsConfig Disabled()
```
Configuring TLS is much easier now. No more combinations of boolean flags and exceptions for unclear states. Just an enum of three options.

flyingsilverfin replied:
great

farost commented on csharp/Test/Behaviour/Query/QuerySteps.cs:
```diff
@@ -124,8 +124,8 @@ private IValue GetRowGetValue(int rowIndex, string conceptKind, string variable,
-                return Tx.Query(queryText, _queryOptions).Resolve();
-            return Tx.Query(queryText).Resolve();
+                return Tx.Query(queryText, _queryOptions).Resolve()!;
+            return Tx.Query(queryText).Resolve()!;
```
Remove warnings

farost commented on http-ts/src/index.ts:
```diff
@@ -131,6 +132,10 @@ export class TypeDBHttpDriver {
         return this.apiGet(`/v1/version`);
     }
 
+    getServers(): Promise<ApiResponse<ServersListResponse>> {
```
New HTTP endpoints that mirror gRPC updates

farost commented on http-ts/tests/behaviour/feature/BUILD:
```diff
@@ -16,34 +16,37 @@
-    data = [],
+)
+
+# Cluster-only test (no core equivalent)
```
Spent quite some time making the encryption work. Not sure if it works 100%, will finish it after the rebases

farost commented on http-ts/tests/behaviour/steps/connection.ts:
```diff
@@ -57,3 +63,58 @@ Then("connection has {int} user(s)", async (expectedUserCount: number) => {
 
 When("connection closes", closeConnection);
+
+// Cluster/Replica steps
```
New steps that can be implemented in HTTP (no server routing, etc)

farost commented on http-ts/tests/behaviour/steps/context.ts:
```diff
@@ -89,12 +93,105 @@ export function setConcurrentAnswers(answers: QueryResponse[]) {
+];
+
+// For cluster mode tests, replace global fetch with one that handles mTLS
+if (isClusterMode) {
```
TODO: ensure it works nicely and clean up if needed

farost commented on rust/src/common/address/address_translation.rs:
```diff
@@ -0,0 +1,44 @@
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) enum AddressTranslation {
+    // public : private
```
Could not decide how to make the interface more explicit (not to confuse public and private)

farost commented on rust/src/common/address/addresses.rs:
```diff
@@ -0,0 +1,232 @@
+
+/// A collection of server addresses used for connection.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Addresses {
```
For ease of use and explicitness, addresses are now an `enum`. Yes, I love enums.

farost commented on rust/src/common/error.rs:
```diff
@@ -128,50 +128,48 @@ error_messages! { ConnectionError
         10: "Received a response with unknown request id '{request_id}'",
     UnexpectedResponse { response: String } =
         11: "Received unexpected response from server: '{response}'. This is either a version compatibility issue or a bug.",
-    InvalidResponseField { name: &'static str } =
```
Decided to break the enumeration like in server. Not sure when it's time to freeze the indices

farost commented on rust/src/common/error.rs:
```diff
@@ -413,45 +448,64 @@ impl From<Status> for Error {
     fn from(status: Status) -> Self {
```
Tried to make it nicer

farost commented on rust/src/connection/network/proto/server.rs:
```diff
@@ -0,0 +1,69 @@
+    fn try_from_proto(proto: ServerProto) -> Result<Server> {
+        let replication_status =
+            proto.replica_status.map(|status| ReplicationStatus::try_from_proto(status)).transpose()?;
+        match proto.address {
```
We receive only one address from proto: connection address!

farost commented on rust/src/connection/network/transmitter/rpc.rs:
```diff
@@ -150,37 +155,15 @@ impl RPCTransmitter {
-                    Some(Err(status)) => Err(status.into()),
-                    None => Err(Error::Connection(ConnectionError::UnexpectedConnectionClose)),
-                }
+                let timeout = rpc.request_timeout();
```
Transaction opening is under the request timeout! Commands within transactions are not affected by request timeout.

farost commented on rust/src/connection/network/channel.rs:
```diff
@@ -54,11 +55,22 @@ pub(super) fn open_callcred_channel(
-        let tls_config =
-            driver_options.tls_config().clone().expect("TLS config object must be set when TLS is enabled");
-        builder = builder.tls_config(tls_config)?;
+    let is_tls_enabled = driver_options.tls_config.is_enabled();
```
An explicit TLS configuration

farost commented on rust/src/connection/server/mod.rs:
```diff
@@ -0,0 +1,192 @@
+
+pub const DEFAULT_SERVER_ID: u64 = 0;
+
+pub trait Replica: Clone {
```
Split servers to `AvailableServer` and `UnavailableServer` for Rust only. `AvailableServer`s can be communicated with, `Unavailable` ones are only for the info. Helps type checking a lot.

farost commented on rust/src/connection/server/server_manager.rs:
A new component that manages everything related to servers, including connections!

farost replied:
Driver becomes just a wrapper for the logic.

farost commented on rust/src/connection/server/server_manager.rs:
```diff
@@ -0,0 +1,488 @@
+impl ServerManager {
+    const PRIMARY_REPLICA_SELECTION_TIMEOUT: Duration = Duration::from_secs(2);
+
+    // TODO: Introduce a timer-based connections update to have a more actualized connections list
```
User-facing methods always do the network calls to receive the most recent info. If it happens, we update our caches. However, the caches can be stale without user calls, and, say, an inactive driver can live for a day, and then start sending request to an outdated primary. It will require some time to find a new primary. Also, the configuration could change drastically. 

Not really needed, but it might be nice to have. I'm currently inclined to remove this, especially with the introduction of hints in the GRPC errors (I don't yet use them here). TODO!

farost commented on rust/src/connection/server/server_manager.rs:
```diff
@@ -0,0 +1,488 @@
+        driver_lang: impl AsRef<str>,
+        driver_version: impl AsRef<str>,
+    ) -> Result<Self> {
+        let (source_connections, replicas) = Self::fetch_servers_from_addresses(
```
We use the addresses passed to the driver only for the initial connection. You can pass a single address from a cluster of 10 -- the driver will fetch and automatically remember all the peers of this single replica, because the replica returns them. It returns the `connection-address`es of the peers, not their serving addresses, so it just... works out of the box!

farost commented on rust/src/connection/server/server_manager.rs:
```diff
@@ -0,0 +1,488 @@
+    }
+
+    #[cfg_attr(feature = "sync", maybe_async::must_be_sync)]
+    pub(crate) async fn execute<F, P, R>(&self, server_routing: ServerRouting, task: F) -> Result<R>
```
Execute based on the server routing

farost commented on rust/src/connection/server/server_manager.rs:
```diff
@@ -0,0 +1,488 @@
+        P: Future<Output = Result<R>>,
+    {
+        match server_routing {
+            ServerRouting::Auto => self.execute_strongly_consistent(task).await,
```
Currently, if the routing is automatic, we execute strongly (by picking the current primary replica).  In the future, we can have other configurations for execution strategies (e.g., find the closest and most easily accessible replica for a read operation)

farost commented on rust/src/connection/server/server_manager.rs:
```diff
@@ -0,0 +1,488 @@
+    {
+        let mut connection_errors = HashMap::new();
+        for replica in replicas.into_iter() {
+            // TODO: We won't ever reconnect to disconnected / new replicas when using execute_on_any.
```
TODO for myself

farost commented on rust/src/connection/server/server_manager.rs:
```diff
@@ -0,0 +1,488 @@
+            .await;
+
+        match result {
+            // TODO: We may not check newly added configured addresses, and we need to somehow integrate it here...
```
TODO for myself

farost commented on rust/src/connection/server/server_manager.rs:
```diff
@@ -0,0 +1,488 @@
+
+    #[cfg_attr(feature = "sync", maybe_async::must_be_sync)]
+    async fn wait_for_primary_replica_selection() {
+        // FIXME: blocking sleep! Can't do agnostic async sleep.
```
Copied from the 2.x implementation.

farost commented on rust/src/connection/driver_options.rs:
````diff
@@ -17,46 +17,77 @@
+/// # Examples
+///
+/// ```rust
+/// let options = DriverOptions::new(DriverTlsConfig::default()).request_timeout(Duration::from_secs(30));
````
New API for `DriverOptions` -- same as for other options in the lib.

farost commented on rust/src/database/database.rs:
```diff
@@ -202,276 +178,4 @@ impl Database {
-    }
-
-    #[cfg_attr(feature = "sync", maybe_async::must_be_sync)]
-    pub(super) async fn run_on_primary_replica<F, P, R>(&self, task: F) -> Result<R>
```
You can see that I've stolen some of the ideas from before. Now, it got reshaped and moved to `server_manager`.

farost commented on rust/src/driver.rs:
```diff
@@ -67,21 +62,17 @@ impl TypeDBDriver {
     #[cfg_attr(
         feature = "sync",
-        doc = "TypeDBDriver::new(\"127.0.0.1:1729\", Credentials::new(\"username\", \"password\"), DriverOptions::new(true, None))"
+        doc = "TypeDBDriver::new(Addresses::try_from_address_str(\"127.0.0.1:1729\").unwrap(), Credentials::new(\"username\", \"password\"), DriverOptions::new(true, None))"
```
New API for parsing addresses

farost commented on tool/test/resources/encryption/ext-grpc-certificate.pem:
They got outdated

farost commented on tool/test/resources/encryption/int-grpc-root-ca.pem:
Can be used for internal clustering protocols of clusters

farost commented on tool/test/start-cluster-servers.sh:
New logic for clustered typedb!

farost committed:
Self PR review

farost commented on RELEASE_TEMPLATE.md:
```diff
@@ -1,3 +1,6 @@
+**This is an alpha release for CLUSTERED TypeDB 3.x. Do not use this as a stable version of TypeDB.**
```
TODO: remove

farost commented on .circleci/windows/csharp/test_deploy_snapshot.bat:
```diff
@@ -26,6 +26,7 @@ powershell -Command "Move-Item -Path bazel-typedb-driver\external\typedb_artifac
 RD /S /Q typedb-all-windows
 powershell -Command "Move-Item -Path typedb-all-windows-* -Destination typedb-all-windows"
 START /B "" typedb-all-windows\typedb server --development-mode.enable=true
+START /B "" typedb-all-windows\typedb server --development-mode.enabled=true --server.http.enabled=false
```
bbuuuuuugggg

flyingsilverfin replied:
why did we have to disable http? Ahh it would clash with the other server? Could change port or no? Doesn't grpc port clash too?

farost replied:
We don't need the HTTP in these tests, and it started failing because of the taken port. Why change the port if we don't need it?

farost committed:
Add failover tests for most of the languages

farost commented on rust/tests/integration/cluster/clustering.rs:
This test will be a `failback` test, working on it in the background! Should be done soon we a major enhancement of the test scripts

farost committed:
Add cluster failover integration tests, add temporary fixes to the server manager

farost committed:
Clean up strong consistency executions

farost committed:
Cleanup todos and fix hashes

flyingsilverfin commented on .factory/automation.yml:
```diff
@@ -49,7 +49,7 @@ build:
         bazel run --config=ci @typedb_dependencies//distribution/artifact:create-netrc
         # TODO: Temporarily build only the 3.0-reimplemented drivers:
-        # bazel build //...
+        # bazel build --config=ci //...
```
i believe we haven't re-enabled C++ yet? does it build? if not, just make sure the list is complete

farost replied:
The list is complete

flyingsilverfin commented on c/src/concept/concept.rs:
```diff
@@ -345,9 +321,9 @@ pub extern "C" fn concept_get_datetime(concept: *const Concept) -> DatetimeInNan
+pub extern "C" fn concept_get_datetime_tz(concept: *const Concept) -> *mut DatetimeAndTimeZone {
     match borrow(concept).try_get_datetime_tz() {
-        Some(value) => DatetimeAndTimeZone::new(&value),
+        Some(value) => release(DatetimeAndTimeZone::new(&value)),
```
memory issue before?

farost replied:
Yeah, risky. `DatetimeAndTimeZone` contains a string and a custom `drop` for it.

flyingsilverfin commented on c/src/server/server_routing.rs:
```diff
@@ -0,0 +1,112 @@
+/// enum.
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub enum ServerRoutingTag {
```
Tag... maybe Type or Method?

farost replied:
Changed to `Type`

flyingsilverfin commented on c/src/server/server_version.rs:
```diff
@@ -0,0 +1,55 @@
+#[repr(C)]
+#[derive(Debug)]
+pub struct ServerVersion {
+    distribution: *mut c_char,
```
do we always (and also need to) hold mut?

farost replied:
We don't **have to**, but the type used in memory management (`release_string`), so that's what we store without casting.

flyingsilverfin commented on c/src/driver.rs:
```diff
@@ -0,0 +1,190 @@
+    driver_options: *const DriverOptions,
+    driver_lang: *const c_char,
+) -> *mut TypeDBDriver {
+    let addresses = iterators_to_map(string_array_view(public_addresses), string_array_view(private_addresses));
```
will this just stop short if the arrays are mismatched, or error?

farost replied:
`zip_eq` asserts both iters have equal lengths, so it will panic just like in 2.x. It's okay for C, the drivers on top of it will probably use more high-level containers that will make freaking this up very hard.

flyingsilverfin commented on csharp/Answer/ConceptRow.cs:
```diff
@@ -87,6 +89,7 @@ public QueryType QueryType
         /// <inheritdoc/>
         public IConcept? GetIndex(long columnIndex)
         {
+            Validator.RequireNonNegative(columnIndex, nameof(columnIndex));
```
nice catch

flyingsilverfin commented on csharp/Api/Server/IServer.cs:
```diff
@@ -0,0 +1,52 @@
+namespace TypeDB.Driver.Api
+{
+    /// <summary>
+    /// The metadata and state of an individual raft server of a driver connection.
```
from the user point of view we generally shouldn't expose implementation (raft) - they only need to know about Clusters!

farost replied:
I think we exposed this in docs in 2.x, but I changed all `raft` to `cluster`.

flyingsilverfin commented on csharp/Api/IDriver.cs:
```diff
@@ -43,31 +40,30 @@ public interface IDriver : IDisposable
 
         /// <summary>
-        /// Opens a transaction to perform read or write queries on the database.
+        /// Opens a communication tunnel (transaction) to the given database on the running TypeDB server.
```
try to avoid new terminology - 'tunnel' isn't something we every use. You can just say 'transaction' here it's fine

flyingsilverfin commented on csharp/Api/ServerRouting.cs:
```diff
@@ -0,0 +1,93 @@
+namespace TypeDB.Driver.Api
+{
+    /// <summary>
+    /// Server routing directive for operations against a distributed server. All driver methods have
```
distributed 'cluster' imo

farost replied:
`distributed server` is `cluster` in the context of the protocol to me. We are a client, and there is a server. It contains multiple `replica`s. We did remove `replica`s in favor of `server`s, so it became a little weird generally, but this is fine in terms of the domain to me.

flyingsilverfin commented on csharp/Test/Deployment/NugetApplicationTest.cs:
```diff
@@ -37,7 +37,7 @@ public static void Main()
             string dbName = "nuget-test-database";
 
-            using (var driver = TypeDB.Driver(TypeDB.DefaultAddress, new Credentials("admin", "password"), new DriverOptions(false, null)))
+            using (var driver = TypeDB.Driver(TypeDB.DefaultAddress, new Credentials("admin", "password"), new DriverOptions(DriverTlsConfig.Disabled())))
```
Much better yea!

flyingsilverfin commented on csharp/Test/Integration/FailoverTest.cs:
```diff
@@ -0,0 +1,262 @@
+        private const int PrimaryPollIntervalSecs = 2;
+        private const int PrimaryFailoverRetries = 5;
+        private static readonly string ClusterServerScript =
+            Environment.GetEnvironmentVariable("CLUSTER_SERVER_SCRIPT")
```
do you use this to start up and shut down? How do you shut down?

farost replied:
In order to make all the integration and behaviour tests in sync, I use this script only for killing and restarting the already running servers. You need to start up all the servers first and then shut them down manually.

flyingsilverfin commented on csharp/TypeDB.cs:
```diff
@@ -17,39 +17,66 @@
+        /// The default address of the TypeDB server.
         /// </summary>
-        public const string DefaultAddress = "localhost:1729";
+        public const string DefaultAddress = "127.0.0.1:1729";
```
Interesting - why the change?

farost replied:
Both work (unless you generate SSL certificates for `127.0.0.1` only), but I want to use explicit loopback/broadcast ips everywhere for the server and the clients to stop confusing `127.0.0.1` and `localhost`. `listen` addresses default to `0.0.0.0`, `advertise` default to `127.0.0.1`. We used to have mixed-up values in different places and sometimes in different drivers.

flyingsilverfin commented on dependencies/typedb/artifacts.bzl:
```diff
@@ -25,20 +25,15 @@ def typedb_artifact():
+    native_artifact_files(
+        name = "typedb_cluster_artifact",
+        group_name = "typedb-cluster-all-{platform}",
+        artifact_name = "typedb-cluster-all-{platform}-{version}.tar.gz", # TODO: Make {ext} instead of tar.gz
```
remember this todo

farost replied:
Never forget

flyingsilverfin reviewed (changes requested):
I've left a set of comments on csharp for now, assuming that these can broadly be applied throughout? I'm not sure i have bandwidth to review everything! Perhaps in the next pass I can review Python since that's the most popular entry point!

farost committed:
Addressing PR review naming and docs

farost committed:
Add cluster failover for http-ts (#864)

## Usage and product changes
Add cluster failover logic for the HTTP-TS driver.
                                                                       
## Implementation
### Initial connection 

The driver uses addresses[0] as its starting server (currentOrigin).
This is the same as master — no behavior change for single-address
users.
                                                    
### Per-request flow (`apiReq`)
1. Try current server — Send the request to currentOrigin. If the server
is unreachable or signin fails, treat it as a connection error (null).
2. Follow 421 redirect — If the server responds with 421 Misdirected
(meaning "I'm not the primary, go talk to X"), the driver parses the
primaryAddress from the response body, switches currentOrigin to that
server, re-authenticates, and retries once.
3. Fallback to other addresses — If the result is still null (connection
error) or non-ok (server error, e.g. during leader election), iterate
through all other configured addresses. For each: try the request,
follow any 421 redirect, return the first successful response. If none
succeed, return the last error.
4. All failed — Return a synthetic HDR1: Cannot connect to any server
error.

Sticky routing: Once the driver discovers a working server (especially
the primary via 421 redirect), currentOrigin stays pointed there.
Subsequent requests go directly — no redundant 421 round-trips until
that server becomes unavailable.

Authentication: One token cached per origin. Cleared on origin switch.
On 401, the driver refreshes the token once and retries. If signin
itself fails on a server (e.g. SRV14 "Not yet initialised" during raft
recovery), that server is skipped and fallback continues.

What's NOT configurable:
- No retry count setting — each address is tried once per request
- No timeout settings for failover (the application-level retry loop
controls that)
- No explicit primary preference — the driver just follows what the
cluster tells it
- The address list is fixed at construction time

farost commented on http-ts/src/index.ts:
```diff
@@ -17,21 +17,25 @@
 } from "./response";
 
 const HTTP_UNAUTHORIZED = 401;
+const HTTP_MISDIRECTED = 421;
```
We use this code instead of direct redirects, so the simplified clients (like `curl`) don't automatically hit the unauthorized endpoint

farost committed:
Update proto and artifacts

farost committed:
Update HTTP naming

farost committed:
Update bdd setup to match integration tests

flyingsilverfin commented on python/tests/behaviour/behave_rule.bzl:
```diff
@@ -32,30 +32,29 @@ def py_behave_test(*, name, background, native_typedb_artifact, steps, feats, de
+        background = "@//python/tests/behaviour/background:core",
         toolchains = ["@rules_python//python:current_py_toolchain"],
-        typedb_port = "1729",
+        typedb_port = "1729", # TODO: Not needed?
```
todos

farost replied:
Thanks

flyingsilverfin commented on python/tests/deployment/test.py:
```diff
@@ -35,7 +35,8 @@ class TestDeployedPythonDriver(TestCase):
         super(TestDeployedPythonDriver, cls).setUpClass()
         global driver
-        driver = TypeDB.driver(TypeDB.DEFAULT_ADDRESS, Credentials("admin", "password"), DriverOptions(is_tls_enabled=False))
+        driver = TypeDB.driver(TypeDB.DEFAULT_ADDRESS, Credentials("admin", "password"),
```
So in the next release, all the users will have to update their driver APIs right

farost replied:
Yes

flyingsilverfin commented on python/tests/integration/cluster/BUILD:
```diff
@@ -15,23 +15,28 @@
+    ],
+    env = {
+        "CLUSTER_SERVER_SCRIPT": "$(rootpath //tool/test:cluster-server.sh)",
+        "ROOT_CA": "$(rootpath //tool/test/resources:encryption/ext-grpc-root-ca.pem)",
```
Cool - are these still valid?? amazing

farost replied:
I had to regenerate the pem files, but they will last for a couple of years again

flyingsilverfin commented on python/typedb/api/answer/query_answer.py:
```diff
@@ -37,8 +37,6 @@ def query_type(self) -> QueryType:
         """
         Retrieves the executed query's type of this ``QueryAnswer``.
 
-        :return:
```
didn't we need these in order for the adoc to get generate right :? dont quite remember - as long as it looks good i dont really care!

farost replied:
Maaaaybe something will break on the website, but I don't see a need for these -- the docs generations look fine. If the website view is broken, I'll get it fixed with no issue

flyingsilverfin commented on python/example.py:
```diff
@@ -8,7 +8,8 @@ class TypeDBExample:
         # Open a driver connection. Specify your parameters if needed
         # The connection will be automatically closed on the "with" block exit
-        with TypeDB.driver(TypeDB.DEFAULT_ADDRESS, Credentials("admin", "password"), DriverOptions(is_tls_enabled=False)) as driver:
+        with TypeDB.driver(TypeDB.DEFAULT_ADDRESS, Credentials("admin", "password"),
```
This should be top-line of the release notes for driver (along with other languages!)

farost replied:
Ok!

flyingsilverfin commented on python/rules.bzl:
```diff
@@ -77,7 +77,7 @@ def native_driver_versioned(python_versions):
             url = "https://github.com/typedb/typedb-driver",
             author = "TypeDB Community",
-            author_email = "community@typedb.com",
+            author_email = "core@typedb.com",
```
keep community in these ones pls - that's our contact email i believe

farost replied:
ohhh it's clearly a bug. I thought I caught this when changing the globals. Sorry and thanks

flyingsilverfin reviewed (approved):
Approved with comments

farost committed:
Update community email

farost committed:
Fix build after rebase

farost committed:
Update deps

farost committed:
Add rules_ts_ext since it was removed from typedb_protocol

farost committed:
Use typedb_cluster artifact

farost committed:
Fix rustfmt

farost committed:
Make artifacts private

farost committed:
Update cluster artifact

farost committed:
Fix docs

farost committed:
Fix http-ts connection, comment out csharp jobs

farost committed:
Fix integration tests bazel setup

farost committed:
Fix feature files refs

farost committed:
Fix java bdd for cluster and update refs

farost committed:
Prepare for release 3.10.0-alpha-1

farost committed:
Fix tag

farost committed:
Fix checkstyle

farost committed:
Update deps

farost committed:
Update MODULE.bazel.lock

farost committed:
Make http-ts connection more resilient

farost committed:
Update deps, cleanup CI/CD todos

farost committed:
Update refs

farost committed:
Update rn
